### DISCUSSION TITLE
Vertical Tabs

### DISCUSSION THREAD

**Discussion by dhulihan** [ghostty-org/ghostty#2549] [Feature Requests, Ideas] [answered]
For those of us with way too many shell sessions open at one time...

Equivalent on iTerm2:

![Image](https://github.com/user-attachments/assets/d857d04a-947c-44b3-8685-b98f9c6767ab)

Configuration example (locations being `Top`, `Bottom`, `Left`):

![Image](https://github.com/user-attachments/assets/caf40ffd-267f-4868-b687-d9389a2d0e25)

**Comment by mitchellh** [comment]
Converted to a discussion so this can be discussed a bit further. 

I'm not a big fan of it. We support bottom tabs on Linux because it is provided by the native GUI toolkit. If Apple supported this natively I'd also be supportive, but this requires a totally different UI element to pull it off (either a list view of some sort or a custom tab view). Generally speaking, Ghostty tries to do as little of that as possible.

I also just want to use this discussion to gauge generally interest.

**Reply by dhulihan** [reply]
Thanks for opening this up to discussion! I hadn't seen any requests for vertical tabs here in gh or discord, so it's good to have a reference point for this decision for others :). 

Vertical tabs are a nice-to-have for me for switching between several open projects at a time (sometimes dozens on bad days), but there are workarounds for this (tmux session switching, etc.). Not a deal breaker by any means! Vertical tabs have been getting a little more popular in browsers (zen, brave), so it is also nice to have familiar layouts across tooling. 

On the other hand, maybe I'm just hoarding too many tabs 😉.

**Reply by narqo** [reply]
_I'm not sure if it is worth opening a separate general discussion about tab location option on macOS, so will put it here._

I run iTerm2 in the full screen mode with tabs bar at the bottom. I find the appeal of it because this visually moves the terminal's output higher a bit on the screen. I feel this better utilises the laptop's screen space.

<img width="1470" alt="Screenshot 2024-12-27 at 19 48 19" src="https://github.com/user-attachments/assets/3f647003-05d9-4a4d-9d4b-cce5e85b4473" />

**Reply by WizardlyMeow** [reply]
> Converted to a discussion so this can be discussed a bit further.
> 
> I'm not a big fan of it. We support bottom tabs on Linux because it is provided by the native GUI toolkit. If Apple supported this natively I'd also be supportive, but this requires a totally different UI element to pull it off (either a list view of some sort or a custom tab view). Generally speaking, Ghostty tries to do as little of that as possible.
> 
> I also just want to use this discussion to gauge generally interest.

I came here to see if bottom tabs are supported, how do I do this on linux?

Nevermind sorry figured it out for anyone who needs it gtk-tabs-location = bottom

**Comment by flextribemartin** [comment]
Recently started using ghostty myself and so far I'm liking it. Decided to give it a few weeks as my primary driver before making a decision.
So far I have been able to implement all the configuration I am used to, from iTerm2, including making the font "fat", which I have just gotten used to. Super cool!

However, the one thing I _cant_ yet do with ghostty, is have tabs on the left side of the window. I'm a big fan of having them on the left, because the "list" view gives me a much better overview, and for whatever reason my mind naturally prefers that viewing order.
Not here to claim one is better than the other, but I'm a supporter of implementing those, seeing how this is a discussion, I thought I'd share :-)

Thanks for the amazing work on the terminal!

**Comment by pluiedev** [comment]
Side tabs are possible on linux via gtk-tabs-position, but we might axe that soon because libadwaita doesn't support it natively (we could hack something together with sidebars, but eh.)

**Reply by feakuru** [reply]
thanks for pointing this out here, turned out to be the option i needed when i stumbled upon this issue in my search, and i completely missed it in the docs. i would thus like to voice support for not axing this feature if possible 😄

**Reply by pluiedev** [reply]
Too late - it's already gone in 1.2. We could try to write a vertical tab implementation ourselves but that would come at a cost of potentially visually clashing with all other libadwaita apps. Sorry.

**Comment by aflat** [comment]
I would prefer tabs on the side for the same reason as the original poster. I tend to have many sessions open, and have about 10-20 tabs running, with various states of windows inside those tabs, used for logical grouping. If I'm working on a cluster of machines, I open a tab, do some window splitting to log into the various machines, run whatever needs to be run, then move onto another tab, while the first set of commands continues to run. This keeps the history of my working in that tab session. 

Since I name my tabs, I know what cluster/whatever I'm doing there. Horizontal tabs don't let me name them much, without running out of room. Plus, like I assume everyone, I have a widescreen monitor, so I want more vertical screen space to for output. Sure 1 row isn't much, but it also serves the same purpose to people who remove the toolbar, titlebar, and tab bar, a few extra rows.

**Reply by pluiedev** [reply]
The thing is that we understand there's interest, but we aren't certain whether that interest outweighs the technical costs. (See Mitchell's comment above.) We appreciate voicing your interest but that doesn't substantially change the status quo.

**Reply by gawie-kellerman** [reply]
You know what...  I hear that everywhere - "the status quo"...  That goes without saying, as I don't pay anything for it.  So it is just an opinion, or a gentle hint/request.  

I think the big take away in this status quo is that there are a "million" terminals to choose from, and switching between them is easy ; while most concentrate on technical behind the scenes stuff you/or specific distro integrations, a terminal  could be setting itself apart (distinguish itself) by simply having reasonable vertical tabs.   That will make it unique at this point...  and hook a few, or even convert a few.   This is not directed towards ghostty specifically.

I will switch to a different working terminal today if it gives me good vertical tabs with keybinds for switching.  But since I don't pay any money, I either should start my own terminal (which I don't have the time or appetite for) or wait and see another  30 years.

**Reply by gawie-kellerman** [reply]
@pluiedev - let me ask you the question differently.  What is the value to a project like ghostty to have more users or participating users to your project? Is there any value of any kind - I ask this question with good intent - as I don't understand what makes you either participate for free, or how you would "monetize" your effort.  I ask this with honest good intent of trying to understand the process / and therefore the status quo

**Reply by pluiedev** [reply]
We're all (unpaid) volunteers here, and we work on what is appealing to us, rather than what is "profitable" or what will appeal to the most amount of people. If none of us seems to be interested in a feature (which appears to be true in this case) then that would stay as-is until either one of us finds it something that we'd like to implement, or an external contributor steps up. 

The value of the user is that, well obviously we'd like to see people enjoying our creation, and they also provide a lot of insight into what we've missed (bugs especially). 

For feature requests like this though, I'd like everyone to keep in mind that we're not wish-granting genies. You can post on feature requests with "+1" or "I want this" however many times you like, but that doesn't change the fact that someone has to implement it. In short, a lot of what I see in requests like these (e.g. #189) is that we get demands and long essays explaining why a feature is core to your experience, and not much else - no path to implementation, no volunteering from external contributors, no suggestions to make it a reality. You can't just pressure someone to make them
magically inclined to do what you want - that's just not how FOSS works.

**Reply by gawie-kellerman** [reply]
Thanks for the response.  I would love to have the time available to me to be in such a position.  Bottom line, you have my opinion for what it is worth. I make very little money, at a horrible 3rd world country exchange rate, but I would definitely consider buying or making a donation to a terminal with vertical tabs-  as I spend a lot of time in it; and cannot find anything that I  like (on Linux) and even 10 times worse on Windows actually.

Thanks for replying

**Reply by pluiedev** [reply]
We have far too many opinions and not enough energy to act on them unfortunately. Appreciate the gesture, though :)

**Reply by gawie-kellerman** [reply]
That is why I am giving you the fact that if I find a vertical tabbed terminal today I will switch to it (if it works).

**Reply by gawie-kellerman** [reply]
@pluiedev  - Sorry - you didn't really answer my original question.  Why are you entertaining questions to us "the users"....  If nobody is willing to listen then why listen in the first place?  What value is a user to you?  I am really not trying to be funny or sarcastic!  What is the aim of the discussions if not for a recommendation or feature proposal?  I have used open source software for very long, and never cared to ask - now I am asking.

I was quiet for the fear of embarking on futility - which I am starting to believe was the right strategy.

**Reply by pluiedev** [reply]
TL;DR we hear you, but it leads nowhere to actually getting the job done. Just because there's support behind an idea doesn't mean we are obliged to implement it. Someone needs to come up with a concrete plan behind the implementation, and this right now is not our top priority, partly due to the fact that this is not a trivial feature on a technical level (see Mitchell's comment above).

**Reply by 00-kat** [reply]
(P.S.: this conversation is very off-topic, please consider moving to a different place, such as [the Ghostty Discord server].)

[the Ghostty Discord server]: https://discord.gg/ghostty

**Comment by gawie-kellerman** [comment]
I really enjoy ghostty - but as mentioned by various users the inability to have vertical tab placement with horizontal text is a shortcoming of the GUI toolkits - which I have at least a little respect for (not much, since these GUIs are 3+ decades old already, and sadly I could do that with Visual Basic 3/Delphi in the 90s already).

That said, I became used to ITerm2 vertical tabs;  and have since looked at at least 30 different terminals (since I don't have a Mac anymore - fried the CPU with C++ cross platform compiles) ; and sadly nothing even remotely functional (Tabby comes close...  but fails on hordes of other things).  The tab situation extends itself to KDE and windows on the freely available software.

I find it strange that I am one of (seemingly) very few people bothered by this, as, even though it seems like a petty issue, I found that visually and from an organize point of view, that vertical tabs absolutely just clicked for me and made my experience not only more efficient but also more pleasant.

I think one reason for this, is that your eyes, on console generally is on the left hand side of the windows most of the time (close to the prompt); meaning for somebody with large monitor(s), your eyes are always close to the tabs.  Another reason is that having lots of tabs are also more easily managed and switched vertically than horizontally (even more so when dragging is implemented)...  especially since it is (a) easier to scroll if needed and (b) easier to read/present as the eye travel vertically is much shorter than horizontally 
And - most of the time monitors are wider (2 times+) rather than taller... meaning you have much more space horizontally rather than vertically (which gives you the opportunity to make tabs wider/more descriptive)

**Comment by deltics** [comment]
I recently started exploring Ghostty and would like to use it more, but the lack of vertical tabs is a major sticking point.  Perhaps it's because I've gotten so used to them in iTerm2, but perhaps not.  I think there is a real advantage, ergonomically, as well as in terms of maximizing vertical space.

With horizontal tabs, the position of the titles of open tabs varies according to how many tabs are open, making the number of tabs unclear, including whether there even ARE any tabs.  I have to examine the top of the window, away from my prompt (which is rarely at the top) to see if there are tabs open and then process the placement of titles to identify how many there are (if any).

With vertical tabs, the visual signal that tabs are open is immediately obvious without looking away from the prompt.  The presence of a tab gutter conveys that tabs are open, and the list of tabs (how many and their titles) is easily found at the top of that gutter.

**Comment by assaf758** [comment]
Coming from tmux to Ghostty tabs, in my opinion, a use case of browser-level number of tabs is less likely.
One thing I do think will help is the ability to have the tab name in proximity to it's number (currently they are on the opposite side of the tab) so knowing the keybind to move to a tab will be trivial.

**Reply by gawie-kellerman** [reply]
I have written near essays about the advantages of vertical tabs over and above the quantity thereof..  in:
1. We have wider screens, making the ability to make the tab wider and more descriptive without negatively influencing the console.
2. With or without tab numbers (Alt-1 etc) which I like, it is much easier to get an idea of where you are with left tabs as the area for looking is columnar.
3. Drag and drop operations even are easier as it is over less space.
4. We read from left to right, and in consoles your eyes are on the left of the screen (at the prompt) most of the time...  so you dont have to jump your eyes a mile away....  I know you can split, but I rarely do that, typing the wrong command in the wrong console can be disastrous, so I split in rare cases.  And sadly, top/bottom tabs even on splits is kind of awkward as it splits the terminal and not the window, meaning you cannot split independently in ghossty/konsole etc to my best experience.
5. Easy to scroll tabs vertically, less so with horizontal

Vertical tabs, my initial reaction years ago was, but why, and after I used ITerm for 3 years I was absolutely sold.  I work ssh over various servers, compiles, dockers etc, at once, so having 15+ tabs is nothing extraordinary. 

The feedback I received from the Ghostty team is that they have much bigger issues to address and that this is not even on the radar, and even more difficult than meets the eye as the GUI toolkit does not support it well  (and since I get it for free I am just too happy to have something decent).  

So for a chuckle; my boss describe to me every other day how AI is going to replace me; so it might be that AI takes over and render us all into a heap before we have a decent vertical tab for a Linux console.

**Reply by pluiedev** [reply]
If you truly feel that strongly about it, PRs are always welcome. In open-source everyone is a volunteer; so be the change you want to see in the world. Just being hoity-toity about it does absolutely nothing for you and I.

**Reply by gawie-kellerman** [reply]
I can answer you in so many ways.   But again, why do you not take on my opinion... and differ from it, or tell me I am wrong;  instead of telling me what to do with my time?  If you don't want it, you can ignore it (like you did up to now) - again - as said, you don't owe me a penny and neither do I, so take it or leave it.  I do have a job that takes 12+ hours of every day, and a family.  So not only do I not have time, but I also don't have the appetite to start or continue a terminal...   Rather take the positive, and that is that I am posting here, meaning that I did take my precious time to raise my silly opinion HERE, meaning that I like what you did with ghostty up to now but would like to see more.   Do I have to explain it to you like for my child, or do you understand?

If this is the kind of response, then why do I bother raising it.  Cheers, last time I talked about this.

**Comment by Xuanwo** [comment]
I want this because it can be really useful for running multiple AI coding agents at the same time.

**Reply by pluiedev** [reply]
We understand the need, but bump comments don't make it any easier or compelling to implement, and only serve to send everyone here a needless notification/email in their inbox. The upvote button exists as a more direct, less disruptive way to do exactly what you have done here. See #6937 for more details.

**Comment by jairo-caro-ag** [comment]
I've implemented a version of vertical tabs with live previews in PR #9931: https://github.com/ghostty-org/ghostty/pull/9931

It adds a new `macos-tab-sidebar` config option that replaces the native tab bar with a sidebar showing live thumbnail previews of all terminals. Features include:

- Live preview thumbnails (~5fps)
- Adaptive grid layout (auto-adjusts columns to fit all tabs)
- Tab number badges (1-9) with Cmd+1-9 keyboard navigation
- Click to select, hover to close
- Context menu support

Would love feedback from anyone interested in this feature!

**Comment by aflat** [comment]
I wasn't looking for a whole mini window as the side tab, I'm more interested in something like 
<img width="1579" height="851" alt="image" src="https://github.com/user-attachments/assets/f1facf14-e865-41bc-862c-c9489f4edb5d" />

Which I did here https://github.com/aflat/ghostty/tree/vert_tabs with almost all AI. Unfortunately I mixed in some changes for linux, which I need to refactor out(I haven't tested it, so I don't want any of that code in) before I even attempt a PR

**Comment by Diaspar4u** [comment]
The issue is that for most of our horizontally rectangular screens, the horizontal space is plentiful while the vertical is very precious - and in vertical stack (like menus on websites, etc) it's easy to glance and switch, especially if running many processes / terminals). so definitely a very high vote for this feature! Not only that, but it'll make it easier to color-code them, rename, etc - since not just using native

**Comment by marcusquinn** [comment]
https://tabby.sh/ does vertical tabs, and more. Really like the way they do profiles, too.

**Comment by gawie-kellerman** [comment]
I used tabby for a while, but it has a bunch of issues that are really annoying, many/most of which is in that (especially on Windows) that parts of the display is not updated.  I also had very annoying keyboard related issues with Tabby.  It could also be much more useful if it simply gave you a slider to set the size of the tabs, as I found the tab area to be too small.  Had a lot of small really annoying other issues as well; but yes, especially on Windows it is a very viable option if the quirks gets fixed or you can live with it.

**Reply by pluiedev** [reply]
Please keep the discussion on topic.

**Comment by gawie-kellerman** [comment]
I really don't know why I waste my time participating...  

How is the discussion not on topic.  Not only was the previous remark deemed on topic 2 replies before, but my response also eludes to an opinion that vertical tabs is no good unless resizable.  

In all honesty, don't bother to reply.  I will make effort to find out how to remove myself from notifications from this thread.

**Reply by pluiedev** [reply]
Because this is the discussion board for Ghostty and not Tabby? The hostility and overall immaturity do not help your case at all either. 

Honestly, if your behavior makes you unable to contribute meaningfully and respectfully to the conversation at all, I will have to consider forcibly removing you from the discussion to avoid further pestering others subscribed here who are expecting productive discussions instead of noise. This is your last warning.

**Reply by 00-kat** [reply]
> Not only was the previous remark deemed on topic 2 replies before

The link to Tabby was almost certainly intended as *prior art*, and not as a recommendation to all users in this thread to switch from Ghostty to Tabby. Even if that was not the intention, it wasn't considered off-topic by us because it still served the purpose of bringing up prior art. However, your comment barely contributed to the discussion viz. feature design for incorporating *vertical tabs* in *Ghostty*—the main thing of note from your comment is that a resizable tab area is greatly preferable; discourse about rendering and input issues, while concerns in Tabby, aren't useful in a discussion about vertical tabs in Ghostty.

It is worth noting that a comment about Tabby does not change the topic, even for subcomments: discussion about Tabby's approach considering relevance to Ghostty would be on-topic, but general discussion on Tabby would not, even if it would be fine if the comment about Tabby was taken out of context.

Even if it could be argued that the presence of one useful tidbit makes the whole comment on-topic, we would still mark it as off-topic as Leah did to avoid encouraging other repliers who will undoubtedly go even further off-topic.

Finally, if you actually want these issues with Tabby fixed, saying they're annoying here will do absolutely nothing; consider bringing up your issues with Tabby's maintainers or 👍️ an existing issue on their issue tracker if present.

**Comment by RomanMIzulin** [comment]
bump. Modern display resolutions are better suited for vertical tabs. Even firefox have added vertical tabs

**Comment by aflat** [comment]
@pluiedev I have a branch ready to do a PR with, https://github.com/aflat/ghostty/tree/mac-vert-tabs but docs say to have an active issue. The issue for this was converted to this discussion. Is it ok to open a PR with this, or should I create a new issue to create the PR?

**Reply by pluiedev** [reply]
I'm not a macOS dev and as such I can't really evaluate whether your approach is correct or not, but one thing to keep in mind is that you should NOT make changes you cannot test (which in this case means you must not change `gtk-tabs-location`). At present your branch simply won't compile for the GTK side of the story.

**Reply by aflat** [reply]
Darn, good catch, I tried to take all those out, missed one. I'll make sure to take that out before I submit the PR. Do you know if it's ok to submit the PR based on the discussion, or do I need to open a new issue?

**Reply by aflat** [reply]
For clarification, I did factor out the gtk changes, and repushed to the same branch

**Reply by aflat** [reply]
@pluiedev is this enough to get an issue created?

**Reply by pluiedev** [reply]
Again, not a macOS dev, can't vouch for the quality of this. If any macOS maintainer thinks this looks okay, then we'll allow this as a PR.

**Comment by kirhgoff** [comment]
Would love to have vertical tabs (and preferrably on the right side too), it is so much more ergonomic, use it in browser, miss it in Ghostty.

**Comment by dasch** [comment]
Just chiming in: with the whole LLM coding revolution going on, I'm also finding myself having an ungodly number of open tabs. I think a sidebar approach offers benefits above and beyond supporting larger numbers of open panes, though – it can also provide a better user experience, although this may run counter to the project goals.

I'm specifically thinking about a "repos and worktrees" approach to managing concurrent work; I'll have a handful of open projects, which will each have their own repo, but within each I'll have 1-10 concurrent agents, each running in their own git worktree within that repo. Right now my head explodes on a daily basis; I would _love_ a MacOS native UI for managing this: a sidebar in Ghostty with level 1 entries being repo dirs, and level 2 being the relative path of the worktree, something like:

```
~/Code/project-1
  .worktrees/upgrade-foo
  .worktrees/refactor-bar

~/Code/project-2
  .worktrees/get-rid-of-x

etc etc
```

This would not be the "tabs on the side" view, but a complementary one. This may be entirely out of scope, though.

**Reply by rhodes-b** [reply]
this is an entirely separate request that already exists (its slightly different but accomplishes the same thing without limiting it to a pwd and subdir)
https://github.com/ghostty-org/ghostty/discussions/9179

**Comment by avisser** [comment]
On HackerNews today, "Ghostty-based macOS terminal with **vertical tabs** and notifications for AI coding agents"

https://github.com/manaflow-ai/cmux

**Reply by CanRau** [reply]
Just stumbled upon it, also got cmd+f search 🥰

**Comment by 8bittts** [comment]
i added vertical tabs but they aren't persistent as you can hide them. simple and [out of the way](https://blog.yen.chat/p/your-terminal-has-a-sidebar-now):

![0a69a418-1492-4e78-b17a-ef0b7e1fd4da_2790x1686](https://github.com/user-attachments/assets/c94ac3e9-0e0b-467f-906d-373bbc8c8228)

**Comment by tomreinert** [comment]
I switched from ghostty to cmux because of the sidebar, but cmux was too unstable for my taste.
So back to ghostty, and I made a fork with an opinionated sidebar. I realized this is more then "Tabs are on the side", so its probably not an easy feature to implement into ghostty in a way it doesn't obstruct core functionality and keeps things simple. 

However, if this is useful, have a look:

https://github.com/tomreinert/ghostty/

<img width="1125" height="749" alt="ghostty-sidebar" src="https://github.com/user-attachments/assets/53d6ac86-0bdb-4e1b-aa37-0e9981f9fbbf" />

**Reply by dasch** [reply]
That's basically what I'm looking for! Also tried cmux and found it too unstable, as well.

**Reply by 8bittts** [reply]
may i ask what made it feel "unstable"?

**Reply by tomreinert** [reply]
> may i ask what made it feel "unstable"?

Little things like Cmd+tab not working sometimes, tabs moving to new windows suddenly.

**Reply by assaf758** [reply]
Very nice!!
Improvement suggestion - add ability to create tabs from ghosttyctl, paving the way for simple tmuxinator-style tabs-restore script.

**Reply by tomreinert** [reply]
> That's basically what I'm looking for! Also tried cmux and found it too unstable, as well.  

If you want to give it a try, let me know how it works for you!

**Comment by hellosagar** [comment]
I want to add a slightly different perspective here that goes beyond vertical tabs as a UI preference.

I've been using cmux as my daily driver recently (alongside Ghostty) and the thing that's changed my workflow isn't just "tabs on the side" — it's the project-level workspace hierarchy. In cmux, the sidebar isn't just a vertical tab list. Each entry is a workspace tied to a project directory, and within each workspace you have your own set of tabs and splits. So I might have:

- **Mobile  App** → Claude Code session, dev server, git terminal
- **Mobile Scripts** → separate Claude Code instance, separate shell
- **Sandbox** → throwaway experiments

Jumping between these with `Cmd+1-8` and then switching tabs within a workspace with `Ctrl+1-8` gives you a two-level hierarchy that flat tabs (horizontal or vertical) just can't replicate.

The reason this matters more now than it did a year ago is agentic coding. When you're running Claude Code or Codex sessions across multiple projects in parallel, each project needs its own cluster of terminals. With flat tabs you end up with 15+ tabs and no way to tell which ones belong together. @dasch's comment about the repos and worktrees approach is exactly the same pain point.

That said, I totally understand the Ghostty philosophy of sticking to native platform UI and not building custom chrome. So rather than asking for a built-in sidebar, I'm curious whether the scriptability and tmux control mode work planned for 1.4 could enable this kind of workspace layer to be built on top of Ghostty maybe through something like `ghosttyctl` or the scripting API. That way Ghostty stays lean and opinionated, but power users who need project-level organization can build or install it as an add-on.

The fact that cmux, @tomreinert's fork, @aflat's branch, @jairo-caro-ag's PR, and @8bittts' implementation all independently arrived at "sidebar with workspace organization" suggests there's a real pattern here. Would love to hear if the 1.4 scripting work could make this possible without needing a fork.

**Reply by rhodes-b** [reply]
Like I have already noted above that is a entirely separate feature request than vertical tabs and exists as its own discussion please stop requesting it in vertical tabs. In fact its even better because its not limited to a specific "project" (whatever a individual calls a project) its whatever arbitrary grouping you want
https://github.com/ghostty-org/ghostty/discussions/2549#discussioncomment-15885185

> With flat tabs you end up with 15+ tabs and no way to tell which ones belong together

this is also already solvable today you can use separate windows to discern what belongs to a specific project if that is really critical and or a tab with many splits where you just zoom between each of them

**Reply by marcusquinn** [reply]
@hellosagar I appreciate your contribution

**Reply by 8bittts** [reply]
I, of course, like all of the implementations and there's a lot of energy behind this...

... but counter-intuitively I think it should stay out of core. Something I appreciate that @mitchellh has talked about often is deciding what not to include since feature bloat is so easy.

I like all of this convo but it feels like this adding any type of fundamental "workspace" is too far off the path of the OG terminal.

I'm not sure I can tell you more than just a feeling / intuition. The original terminal was feature thin and certainly don't have much workspace features.

I might be shooting myself in the face here though. I am known to do that.

**Reply by marcusquinn** [reply]
Maybe, and they can be as minimalist as they want. The hyper-productive ain't using a demo minimalist app that doesn't have vertical tabs. Wrong time to be going slow. It's barely even a feature request, more of a basic courtesy to the host OS.

**Reply by tomreinert** [reply]
I think the problem here is that people have different mental models how to organize their spaces.
For example I didn't like the nested tabs in cmux. In my head it's a terminal window for each "project" and tabs for each task. But thats just me. I don't even use splits, so maybe I'm not the standard user.

During implemenation I noticed that sidebar is a rabbit hole, really hard to make a one-size-fits-all solution and there are so many implications.

So I agree that this probably shouldn't be part of core. Or be part of it but as close to standard OS behaviour as possible.

**Reply by marcusquinn** [reply]
As an observer, it seems more a problem of minimalist ego versus over-complicating the already well-worn and used standard — resulting in failure from not trying.

I just use Tabby, because it provides for all the basic necessities, and makes me more productive over dozens of projects.

I offer moral support to this basic feature here — and the same to the Warp repo — because it's such an obvious omission.

Technically, most browsers and apps have this now, so there's nothing to invent, just implement. No need for analysis paralysis.

If maintainers don't care for it, I just take that as a sign of ego or saturated attention — and flush this app out of further consideration, accordingly, dismissing it (in my opinion) as a fad project.

I admire the energy to have even started a project like this but, as an observer, these convoluted debates tell me not to rely on it — and just see it as an MVP for other's to then fork or take inspiration from, and do better with.

Which, I expect, is why cmux and tabby exist to fill these gaps.

Kudos to those persevering with proposals but, for me, the lack of interest in making this nominal and standard feature becomes a sign to move on from.

**Comment by jcollie** [comment]
Locking this conversation. Nothing _useful_ has been said in a _long_ time. We know it's a feature that people desire but it's not something that's going to be developed any time soon given the practical realities. Everything else has just been noise.

**Comment by mitchellh** [comment, marked as answer]
I'll do one better than what @jcollie added, I'm going to close this issue. Vertical tabs requires -- at least on macOS -- custom tab bars. We do plan on implementing custom tab bars at some point because there's quite a lot of functionality (particularly on macOS) that the native tab bar just can't get us. 

However, its so far and and so uncertain that I don't see any need to keep this feature request open when the short term answer is "no." Ghostty forks enable vertical tabs and thats an excellent option if you want that.
